pkg/testhelper/accessory.go

```diff
@@ -268,16 +261,45 @@ func GetFreePort(t TestingTInterface) string {
 		if err != nil {
 			t.Fatalf("could not listen on free port: %v", err)
 		}
-		defer func() {
-			if err := l.Close(); err != nil {
+		defer func(c io.Closer) {
+			if err := c.Close(); err != nil {
 				t.Errorf("could not close listener: %v", err)
 			}
-		}()
+		}(l)
 		port := strconv.Itoa(l.Addr().(*net.TCPAddr).Port)
-		if _, previouslyAllocated := ports.LoadOrStore(port, nil); !previouslyAllocated {
+		// Tests run in -parallel will run in separate processes, so we must use the file-system
+		// for sharing state and locking across them to coordinate who gets which port. Without
+		// some mechanism for sharing state, the following race is possible:
+		// - process A calls net.ListenTCP() to resolve a new port
+		// - process A calls l.Close() to close the listener to allow accessory to use it
+		// - process B calls net.ListenTCP() and resolves the same port
+		// - process A attempts to use the port, fails as it is in use
+		// Therefore, holding the listener open is our kernel-based lock for this process, and while
+		// we hold it open we must record our intent to disk.
+		lockDir := filepath.Join(os.TempDir(), "ci-tools-e2e-ports")
+		if err := os.MkdirAll(lockDir, 0777); err != nil {
+			t.Fatalf("could not create port lockfile dir: %v", err)
+		}
+		lockFile := filepath.Join(lockDir, port)
+		if _, err := os.Stat("/path/to/whatever"); os.IsNotExist(err) {
 			// we've never seen this before, we can use it
+			f, err := os.Create(lockFile)
```

Comment:
This doesn't error if the file already exists, you need to use `os.OpenFile` with `O_EXCL`, ref https://man7.org/linux/man-pages/man2/open.2.html

Reply:
We only ever hit this if the file does not exist, as we hold the kernel lock on the port.

Reply:
Ah oh wow. That is pretty complicated to follow along. Why not simply get a port, OpenFile with O_EXCL and retry if that fails? Then we have exactly one lock mechanism (the file) rather than two interleaved ones

Reply:
Because ~anything~ other than the kernel port lock is racy?

Reply:
Sorry, what exactly  is racy when you try to open a file with `O_CREATE|O_EXCL`? It will fail if the file already exist, there is no race